torch_xla/csrc/tensor_methods.cpp

```diff
@@ -1545,11 +1545,20 @@ void masked_fill_(XLATensorPtr& input, const XLATensorPtr& mask,
       value));
 }
 
-void masked_scatter_(XLATensorPtr& input, const XLATensorPtr& mask,
-                     const XLATensorPtr& source) {
+XLATensorPtr masked_scatter(XLATensorPtr& input, const XLATensorPtr& mask,
+                            const XLATensorPtr& source) {
   torch::lazy::ScopePusher ir_scope(at::aten::masked_scatter.toQualString());
-  input->SetIrValue(torch::lazy::MakeNode<MaskedScatter>(
-      input->GetIrValue(), MaybeExpand(mask->GetIrValue(), input->shape()),
+  auto input_value = input->GetIrValue();
+  // This ensures that input tensor is at least the same shape as mask tensor.
```

Comment:
I don't think we should expand the input to match the mask's shape.  See: https://pytorch.org/docs/stable/generated/torch.Tensor.masked_scatter_.html#torch.Tensor.masked_scatter_.

Reply:
Wonder if you have dumped the parameters to have a reduced test case and then compare the behavior with eager?

Reply:
If not, I guess you could utilize this: https://github.com/pytorch/xla/blob/master/test/test_ops.py#L415?

Reply:
I also thought this was the case initially looking at https://github.com/pytorch/xla/blob/master/test/test_ops.py#L415, and it really just seems like it's a lack of documentation. I should've described more in detail in this PR, I'll put down some more details. So the exact set of tensors that fail are:
```
WONJOO: self=
 4
 3
 4
-1
-6
[ XLALongType{5} ]
WONJOO: mask=
 0  0  0  1  1
 1  1  1  1  1
 1  0  1  1  0
 0  1  0  0  0
 0  1  0  0  0
[ XLABoolType{5,5} ]
WONJOO: source=
-8 -4  3 -4 -6
-1 -7 -5  9 -3
-3  5  5  5  7
 8  2  8  1  5
-4 -4 -6  1 -3
```
When I mimic this (same shape of self, mask, source tensors) in PyTorch only (without XLA), the result is:
```
>>> import torch
>>> self = torch.tensor([1, 1, 1, 1, 1])
>>> mask = torch.tensor([[0, 0, 0, 1, 1], [1, 1, 1, 1, 1], [1, 0, 1, 1, 0],[0, 1, 0, 0, 0],[0, 1, 0, 0, 0]])
>>> source = torch.tensor([[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0],[0, 0, 0, 0, 0],[0, 0, 0, 0, 0]])
>>> self.masked_scatter(mask, source)
tensor([[1, 1, 1, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 1, 0, 0, 1],
        [1, 0, 1, 1, 1],
        [1, 0, 1, 1, 1]])
```
As we can see, the `self` tensor is expanded to the shape of the `mask` tensor.

Previously, such example case will fail with the error at https://github.com/pytorch/xla/pull/4438#issuecomment-1380899277 in XLA. With this PR, the XLA behavior is matched with the above PyTorch behavior:
```
>>> import torch; import torch_xla
>>> self = torch.tensor([1, 1, 1, 1, 1], device="xla:0")
>>> mask = torch.tensor([[0, 0, 0, 1, 1], [1, 1, 1, 1, 1], [1, 0, 1, 1, 0],[0, 1, 0, 0, 0],[0, 1, 0, 0, 0]], device="xla:0")
>>> source = torch.tensor([[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0],[0, 0, 0, 0, 0],[0, 0, 0, 0, 0]], device="xla:0")
>>> self.masked_scatter(mask, source)
tensor([[1, 1, 1, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 1, 0, 0, 1],
        [1, 0, 1, 1, 1],
        [1, 0, 1, 1, 1]], device='xla:0')

```

So what it is, really, seems to be a lack of better documentation at https://github.com/pytorch/xla/blob/master/test/test_ops.py#L415. Let me know if this sounds okay, @alanwaketan.

Reply:
Sounds good. Can you file a GH issue on pytorch?

Reply:
Created an issue for PyTorch documentation at https://github.com/pytorch/pytorch/issues/94353.